runtime/testy/Ansible.py

```diff
@@ -223,15 +230,99 @@ def start(self):
         try:
             while True:
                 nextCall = time.time()
+                nextCall += 1.0/UDPRecvClass.UDP_RECV_HZ
                 event = sel.select()
                 self.udpReceiver()
                 self.unpackageData()
                 try:
-                    time.sleep(nextCall - time.time())
+                    time.sleep(nextCall +  - time.time())
                 except ValueError:
                     continue
         except Exception as e:
             self.badThingsQueue.put(BadThing(sys.exc_info(),
             "UDP receiver thread has crashed with error: " + str(e),
             event = BAD_EVENTS.UDP_RECV_ERROR,
             printStackTrace = True))
+
+class TCPClass(AnsibleHandler):
+    PORT = 1234
+    TCP_HZ = 25
+
+    def __init__(self, badThingsQueue, stateQueue, pipe):
+        self.sendBuffer = TwoBuffer()
+        self.recvBuffer = TwoBuffer()
+        sendName = THREAD_NAMES.TCP_SENDER
+        recvName = THREAD_NAMES.TCP_RECEIVER
+        super().__init__(sendName, TCPClass.sender, recvName, TCPClass.receiver, badThingsQueue, stateQueue, pipe)
+
+        stateQueue.put([SM_COMMANDS.SEND_ADDR, [PROCESS_NAMES.TCP_PROCESS]])
+        self.dawn_ip = pipe.recv()[0]
+
+        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.sock.connect((self.dawn_ip, TCPClass.PORT))
+
+    def sender(self, badThingsQueue, stateQueue, pipe):
+        def packageMessage(data):
+            try:
+                proto_message = notification_pb2.Notification()
+                proto_message.header = notification_pb2.Notification.CONSOLE_LOGGING
+                proto_message.console_output = data
+                return proto_message.SerializeToString()
+            except Exception as e:
+                badThingsQueue.put(BadThing(sys.exc_info(), 
+                    "TCP packager crashed with error: " + str(e),
+                    event = BAD_EVENTS.TCP_ERROR,
+                    printStackTrace = True))
+        def packageConfirm(confirm):
+            try:
+                proto_message = notification_pb2.Notification()
+                if confirm:
+                    proto_message.header = notification_pb2.Notification.STUDENT_RECEIVED
+                else:
+                    proto_message.header = notification_pb2.Notification.STUDENT_NOT_RECEIVED
+                return proto_message.SerializeToString()
+            except Exception as e:
+                badThingsQueue.put(BadThing(sys.exc_info(), 
+                    "TCP packager crashed with error: " + str(e),
+                    event = BAD_EVENTS.TCP_ERROR,
+                    printStackTrace = True))
+        while True:
+            try:
+                rawMessage = pipe.recv()
+                nextCall = time.time()
```

Comment:
The problem with throttling the sending liek this is that if whatever is on the other end of the `pipe` is generating data faster than this thread processes, its gonna fill up and cause a constant latency, as everything on the other end starts blocking on this function. 

Or, if we use non-blocking, the stuff on the other end will error out with `pipe full` errors. 
Especially since this is TCP, it doesn't seem like we want to drop any of these packets. 

It looks like this is purely used for 
1) confirming student code uploaded
2) sending over console output

The former is low frequency/ bandwidth and likely won't cause an issue.
The latter is potentially high frequency, high bandwidth, and most likely WILL cause an issue. Especially if you're throttling.

Maybe batching messages in the pipe together and sending them that way? 


In the interest of getting this merged though, I'm also okay with merging this.
But the console logging commit better fix it.